From 61761f5decc1f7548ab1ebb9899d154054f3b372 Mon Sep 17 00:00:00 2001
From: Peter Stephenson <pws@users.sourceforge.net>
Date: Sat, 3 Dec 2011 17:24:45 +0000
Subject: [PATCH] 29934: Stef van Vlierberghe: uninitialised memory after lexer
 realloc

* From Stef VAN VLIERBERGHE: 29934: Src/lex.c (add): use of
uninitialised memoryx when lexer needed to reallocate token.

Upstream-commit: 724fd07a67f135c74eba57e9f25fd342201ec722
Signed-off-by: Kamil Dudka <kdudka@redhat.com>
---
 Src/lex.c | 12 ++----------
 1 file changed, 2 insertions(+), 10 deletions(-)

diff --git a/Src/lex.c b/Src/lex.c
index 157ca3f..269f491 100644
--- a/Src/lex.c
+++ b/Src/lex.c
@@ -571,22 +571,14 @@ add(int c)
 {
     *bptr++ = c;
     if (bsiz == ++len) {
-#if 0
-	int newbsiz;
-
-	newbsiz = bsiz * 8;
-	while (newbsiz < inbufct)
-	    newbsiz *= 2;
-	bptr = len + (tokstr = (char *)hrealloc(tokstr, bsiz, newbsiz));
-	bsiz = newbsiz;
-#endif
-
 	int newbsiz = bsiz * 2;
 
 	if (newbsiz > inbufct && inbufct > bsiz)
 	    newbsiz = inbufct;
 
 	bptr = len + (tokstr = (char *)hrealloc(tokstr, bsiz, newbsiz));
+	/* len == bsiz, so bptr is at the start of newly allocated memory */
+	memset(bptr, 0, newbsiz - bsiz);
 	bsiz = newbsiz;
     }
 }
-- 
2.1.0

